niworkflows/interfaces/tests/test_bids.py

```diff
@@ -16,9 +17,134 @@
+    (T1W_PATH, ["tfm.txt"], {"from": "fsnative", "to": "T1w", "suffix": "xfm"},
+        "sub-100185/anat/sub-100185_from-fsnative_to-T1w_mode-image_xfm.txt"),
+    (T1W_PATH, ["tfm.h5"], {"from": "MNI152NLin2009cAsym", "to": "T1w", "suffix": "xfm"},
+        "sub-100185/anat/sub-100185_from-MNI152NLin2009cAsym_to-T1w_mode-image_xfm.h5"),
+    (T1W_PATH, ["anat.nii.gz"], {"desc": "brain", "suffix": "mask"},
+        "sub-100185/anat/sub-100185_desc-brain_mask.nii.gz"),
+    (T1W_PATH, ["anat.nii.gz"], {"desc": "brain", "suffix": "mask", "space": "MNI"},
+        "sub-100185/anat/sub-100185_space-MNI_desc-brain_mask.nii.gz"),
+    (T1W_PATH, ["anat.surf.gii"], {"suffix": "pial", "hemi": "L"},
+        "sub-100185/anat/sub-100185_hemi-L_pial.surf.gii"),
+    (T1W_PATH, ["aseg.nii", "aparc.nii"], {"desc": ["aseg", "aparcaseg"], "suffix": "dseg"},
+        [f"sub-100185/anat/sub-100185_desc-{s}_dseg.nii" for s in ("aseg", "aparcaseg")]),
+    (T1W_PATH, ["anat.nii", "anat.json"], {"desc": "preproc"},
+        [f"sub-100185/anat/sub-100185_desc-preproc_T1w.{ext}" for ext in ("nii", "json")]),
+    (T1W_PATH, ["anat.nii.gz"] * 3, {"label": ["GM", "WM", "CSF"], "suffix": "probseg"},
+        [f"sub-100185/anat/sub-100185_label-{l}_probseg.nii.gz" for l in ("GM", "WM", "CSF")]),
+
+    # BOLD data
+    (BOLD_PATH, ["aroma.csv"], {"suffix": "AROMAnoiseICs"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_AROMAnoiseICs.csv"),
+    (BOLD_PATH, ["confounds.tsv"], {"suffix": "regressors", "desc": "confounds"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_desc-confounds_regressors.tsv"),
+    (BOLD_PATH, ["mixing.tsv"], {"suffix": "mixing", "desc": "MELODIC"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_desc-MELODIC_mixing.tsv"),
+    (BOLD_PATH, ["lh.func.gii"], {"space": "fsaverage", "density": "10k", "hemi": "L"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_"
+        "space-fsaverage_den-10k_hemi-L_bold.func.gii"),
+    (BOLD_PATH, ["hcp.dtseries.nii"], {"space": "fsLR", "density": "91k"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_"
+        "space-fsLR_den-91k_bold.dtseries.nii"),
+    (BOLD_PATH, ["ref.nii"], {"space": "MNI", "suffix": "boldref"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_space-MNI_boldref.nii"),
+    (BOLD_PATH, ["dseg.nii"], {"space": "MNI", "suffix": "dseg", "desc": "aseg"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_space-MNI_desc-aseg_dseg.nii"),
+    (BOLD_PATH, ["mask.nii"], {"space": "MNI", "suffix": "mask", "desc": "brain"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_space-MNI_desc-brain_mask.nii"),
+    (BOLD_PATH, ["bold.nii"], {"space": "MNI", "desc": "preproc"},
+        "sub-100185/func/sub-100185_task-machinegame_run-1_space-MNI_desc-preproc_bold.nii"),
+    # Nondeterministic order - do we really need this to work, or we can stay safe with
```

Comment:
Why is the order nondeterministic?

Reply:
I haven't investigated it in depth (just suffered it). There are two points where this could be non-deterministic: the sorting of the input spec inputs might vary or [something I could've messed up here](https://github.com/bids-standard/pybids/blob/master/bids/layout/writing.py#L301-L339).

Anyways, the combination of descriptors seemed to me a behavior that should be avoided (and which we were not using at all, thanks to MapNodes) and I didn't dig up further. I considered adding a warning or directly fail when entities are attempted to expand.

Reply:
Okay. I don't think I really understand what's going on here, but if we're not actually using this functionality, then I guess that's okay.

Reply:
Another potential reason is the way we ``add_traits``. Since we are manipulating sets, the ordering might change between runs - I guess setting the python hash seed would help. But yes, this is here just hinting that it can do the filename generation for you, and maybe in the future we will need to revise. But at this point is not used anywhere (and there's no need for it).